RDKDEV-1093: improve gfx memory handling

tomasz-karczewski-red opened the pull request:
(1) calling 'reshape(0,0)' when destroying webgl context

This releases the textures attached to context fbo faster, without waiting for GC. These textures make up for a big chunk of gfx memory used, and when opening different apps quickly we were running out of gfx memory before GC was able to kick in.

(2) changing order of operations in ~WebGLRenderingContextBase

Removing from the context group before the context itself is removed. Otherwise, WebGLContextGroup will not be able to release some of the 'group shared' resources (these that still have attachements when being removed via WebGLSharedObject destructor) - it happens eg. for WebGLVertexArrayObjectBase objects that have buffers set on destruction time.

tomasz-karczewski-red committed:
webgl context objects leak fix

Changing order of operations in ~WebGLRenderingContextBase
Otherwise, WebGLContextGroup will not be able to release some of the
'group shared' resources that still have attachements when being removed
via WebGLSharedObject destructor - it happens eg. for
WebGLVertexArrayObjectBase objects that have buffers set on
destruction time.

The current behaviour in ~WebGLRenderingContextBase:

- first, the reference to some WebGLSharedObject (like m_boundArrayBuffer)
   is nullified
- since this decreases refcnt to 0, destructor for this WebGLSharedObject is
  executed - this goes via WebGLObject::runDestructor() -> WebGLObject::deleteObject.
- WebGLObject::deleteObject checks for attachement count (m_attachmentCount)
  and if it's not 0 it will not execute deleteObjectImpl (where the actual
  removal from the context takes place).
- So the reference to the object is destructed, and ~WebGLSharedObject removes
  the object from m_contextGroup as well. At this point the resource is leaked.

With the fix, m_contextGroup->removeContext(*this) is executed at the beginning
of ~WebGLRenderingContextBase - this calls detach() on all the objects
(WebGLContextGroup::detachAndRemoveAllObjects) so is possible to release them.

magomez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1262,7 +1269,6 @@ WebGLRenderingContextBase::~WebGLRenderingContextBase()
         detachAndRemoveAllObjects();
         loseExtensions(LostContextMode::RealLostContext);
         destroyGraphicsContextGL();
-        m_contextGroup->removeContext(*this);
```
why are you moving this to the beginning of the function instead of moving it just before the previous line where the garphicsContexGL is destroyed? Wouldn't that work for some reason that I'm missing?

tomasz-karczewski-red replied:
Hello @magomez ,

This context removal from group ('m_contextGroup->removeContext(*this)') must happen before the destructor of given WebGLSharedObject is called. Otherwise what happens in ~WebGLRenderingContextBase is like this:

- first, the reference is nullified ('m_boundArrayBuffer = nullptr' etc.); these are RefPtr's and this decreases the refcnt to this WebGLSharedObject to 0
- so, the destructor for this WebGLSharedObject is executed - this goes via WebGLObject::runDestructor() -> WebGLObject::deleteObject. WebGLObject::deleteObject checks for attachement count (m_attachmentCount) - and if it's not 0 it will not execute deleteObjectImpl (where the actual 'physical' removal from the context takes place). So the reference to the object is destructed, and ~WebGLSharedObject also removes the object from m_contextGroup. At this point, the resource is leaked I believe.

So only moving 'm_contextGroup->removeContext(*this)' before destroyGraphicsContextGL() isn't enough; it needs to happen before these destructors execute. With the submitted fix, m_contextGroup->removeContext(*this) is executed at the beginning of ~WebGLRenderingContextBase - and this calls detach() on everything (WebGLContextGroup::detachAndRemoveAllObjects) so is possible to release all the objects.

Please let me know if that makes sense now (suppose I could explain this better in the commit message?)

Regards,
Tomasz

magomez replied:
Ok, that makes sense. Please split this commit into 2 different ones, because its doing two different things: one fixes the leak and the other is a trick to release memory faster, but it's not related. After that it should be ok to merge.

tomasz-karczewski-red committed:
calling 'reshape(0,0)' when destroying webgl context

This releases the textures attached to context fbo faster, without waiting for GC.
These textures make up for the most of gfx memory used, and when opening different apps
quickly we may run out of gfx memory before GC kicks in.
